drivers/gpio/gpio_mcp23s17.c

```diff
@@ -0,0 +1,520 @@
+				u32_t pin, int flags)
+{
+	struct mcp23s17_drv_data *const drv_data =
+	    (struct mcp23s17_drv_data * const)dev->driver_data;
+	int ret;
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	ret = setup_pin_dir(dev, access_op, pin, flags);
+	if (ret) {
+		LOG_ERR("MCP23S17: error setting pin direction (%d)", ret);
+		goto done;
+	}
+
+	ret = setup_pin_pullupdown(dev, access_op, pin, flags);
+	if (ret) {
+		LOG_ERR("MCP23S17: error setting pin pull up/down (%d)", ret);
+		goto done;
+	}
+
+ done:
+	k_sem_give(&drv_data->lock);
+	return ret;
+}
+
+static int mcp23s17_write(struct device *dev, int access_op,
+			       u32_t pin, u32_t value)
+{
+	struct mcp23s17_drv_data *const drv_data =
+	    (struct mcp23s17_drv_data * const)dev->driver_data;
+	union mcp23s17_port_data *port = &drv_data->reg_cache.gpio;
+	u16_t bit_mask;
+	u16_t new_value;
+	int ret;
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	if (access_op == GPIO_ACCESS_BY_PIN) {
+		bit_mask = BIT(pin);
+
+		new_value = (value << pin) & bit_mask;
+		new_value &= bit_mask;
+
+		port->all &= ~bit_mask;
+		port->all |= new_value;
+	} else { /* GPIO_ACCESS_BY_PORT is not supported */
+		return -ENOTSUP;
+	}
+
+	ret = write_port_regs(dev, REG_GPIO_PORTA, port);
+
+	k_sem_give(&drv_data->lock);
+	return ret;
+}
+
+static int mcp23s17_read(struct device *dev, int access_op,
+			      u32_t pin, u32_t *value)
+{
+	struct mcp23s17_drv_data *const drv_data =
+	    (struct mcp23s17_drv_data * const)dev->driver_data;
+	union mcp23s17_port_data buf;
+	int ret;
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	ret = read_port_regs(dev, REG_GPIO_PORTA, &buf);
+	if (ret != 0) {
+		goto done;
+	}
+
+	if (access_op == GPIO_ACCESS_BY_PIN) {
+		*value = (buf.all >> pin) & 0x01;
+	} else { /* GPIO_ACCESS_BY_PORT is not supported */
+		return -ENOTSUP;
+	}
+
+ done:
+	k_sem_give(&drv_data->lock);
+	return ret;
+}
+
+static int mcp23s17_port_get_raw(struct device *dev, u32_t *value)
+{
+	struct mcp23s17_drv_data *const drv_data =
+	    (struct mcp23s17_drv_data * const)dev->driver_data;
+	union mcp23s17_port_data buf;
+	int ret;
+
+	/* Can't do SPI bus operations from an ISR */
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	ret = read_port_regs(dev, REG_GPIO_PORTA, &buf);
+	if (ret != 0) {
+		goto done;
+	}
+
+	*value = buf.all;
+
+done:
+	k_sem_give(&drv_data->lock);
+	return ret;
+}
+
+static int mcp23s17_port_set_masked_raw(struct device *dev,
+					      u32_t mask, u32_t value)
+{
+	struct mcp23s17_drv_data *const drv_data =
+	    (struct mcp23s17_drv_data * const)dev->driver_data;
+	union mcp23s17_port_data buf;
+	int ret;
+
+	/* Can't do SPI bus operations from an ISR */
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	buf.all = drv_data->reg_cache.gpio.all;
+	buf.all = (buf.all & ~mask) | (mask & value);
+
+	ret = write_port_regs(dev, REG_GPIO_PORTA, &buf);
+
```

Comment:
I may be missing something but it doesn't look like we're updating the cache if writing to the mcp23s17 register is successful.